x/concentrated-liquidity/pool_hooks_test.go

```diff
@@ -0,0 +1,187 @@
+
+			// If expect error on set, check here
+			if tc.expectErrOnSet {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+
+			// Get contract address
+			contractAddress := s.clk.GetPoolHookContract(s.Ctx, 1, tc.actionPrefix)
+
+			// Assertions
+			s.Require().Equal(tc.cosmwasmAddress, contractAddress)
+		})
+	}
+}
+
+// TestCallPoolActionListener tests the high level functionality of CallPoolActionListener,
+// which is the helper that calls CW contracts as part of hooks.
+//
+// Since the function is quite general, we only test basic flows and run a sanity check on the
+// gas limit to ensure it is metered correctly and can't run unboundedly.
+//
+// A basic CW contract is used for testing, which simply counts up to `Amount` and does a state write on each iteration.
+func (s *KeeperTestSuite) TestCallPoolActionListener() {
+	// Skip test if there is system-side incompatibility
+	s.SkipIfWSL()
+
+	tests := map[string]struct {
+		wasmFile      string
+		msg           CountSudoMsg
+		noContractSet bool
+
+		expectedError error
+	}{
+		"valid contract that consumes less than limit": {
+			wasmFile: counterContractPath,
+			msg: CountSudoMsg{
+				Count: CountMsg{
+					// Consumes roughly 100k gas, which should be comfortably under the limit.
+					Amount: 10,
+				},
+			},
+		},
+		"no contract set in state": {
+			wasmFile: counterContractPath,
+			msg: CountSudoMsg{
+				Count: CountMsg{
+					// Consumes roughly 100k gas, which should be comfortably under the limit.
+					Amount: 10,
+				},
+			},
+
+			// We expect this to fail quietly and be a no-op
+			noContractSet: true,
+		},
+		"empty message": {
+			wasmFile: counterContractPath,
+			// We expect this to be a no-op without error
+			msg: CountSudoMsg{},
+		},
+		"error: contract that consumes more than limit": {
+			wasmFile: counterContractPath,
+			msg: CountSudoMsg{
+				Count: CountMsg{
+					// Each loop in the contract consumes on the order of 1k-10k gas,
+					// so this should push consumed gas over the limit.
+					Amount: types.ContractHookGasLimit / 1000,
+				},
+			},
+
+			expectedError: types.ContractHookOutOfGasError{GasLimit: types.ContractHookGasLimit},
+		},
+	}
+	for name, tc := range tests {
+		s.Run(name, func() {
+			s.SetupTest()
+
+			// --- Setup ---
+
+			// Upload and instantiate wasm code
+			cosmwasmAddressBech32 := s.uploadAndInstantiateContract(tc.wasmFile)
+
+			// Set pool hook contract to the newly instantiated contract
+			err := s.clk.SetPoolHookContract(s.Ctx, validPoolId, validActionPrefix, cosmwasmAddressBech32)
+			s.Require().NoError(err)
+
+			// Marshal test case msg to pass into contract
+			msgBz, err := json.Marshal(tc.msg)
+			s.Require().NoError(err)
+
+			// --- System under test ---
+
+			err = s.clk.CallPoolActionListener(s.Ctx, msgBz, validPoolId, validActionPrefix)
+
+			// --- Assertions ---
+
+			if tc.expectedError != nil {
+				s.Require().ErrorAs(err, &tc.expectedError)
```

Comment:
Why not `ErrorIs`? `ErrorAs` only checks error type but not the specific message and fields

Reply:
Thanks for flagging. This was from an earlier version of the error message that tried to include more information like actual gas consumed (see response to your other comment on this for why that isn’t easily doable). Changed to `ErrorIs` here. 46aa7f26684c44b455130981a18944005b331e6c